src/http_workers/work_runner.rs

```diff
@@ -0,0 +1,147 @@
+use std::{sync::Arc, time::SystemTime};
+
+use async_trait::async_trait;
+use chrono::Utc;
+use hyper::{
+    client::{Builder, HttpConnector},
+    Body, Client, Method, Request,
+};
+
+use crate::{
+    cache::Cache,
+    identity::Identity,
+    twin::{SubstrateTwinDB, Twin, TwinDB},
+    types::{Message, QueuedMessage},
+    workers::Work,
+};
+
+use anyhow::{Context, Result};
+
+#[derive(Clone)]
+pub struct WorkRunner<C, I>
+where
+    C: Cache<Twin>,
+    I: Identity,
+{
+    twin_db: SubstrateTwinDB<C>,
+    identity: I,
+}
+
+impl<C, I> WorkRunner<C, I>
+where
+    C: Cache<Twin>,
+    I: Identity,
+{
+    pub fn new(twin_db: SubstrateTwinDB<C>, identity: I) -> Self {
+        Self { twin_db, identity }
+    }
+
+    async fn get_twin(&self, twin_id: usize, retires: usize) -> Option<Twin> {
+        for _ in 0..retires {
+            let twin = self.twin_db.get(twin_id.to_owned() as u32).await;
+
+            match twin {
+                Ok(twin) => {
+                    return Some(twin);
+                }
+                Err(err) => {
+                    log::debug!(
+                        "can not retrieve twin from substrate for this id '{}' because of {}",
+                        twin_id,
+                        err
+                    );
+                    continue;
+                }
+            };
+        }
+        None
+    }
+
+    fn encrypt_dat(dat: String, pubkey: String) -> Result<String> {
+        todo!()
+    }
+
+    async fn send_msg(twin: Twin, uri_path: String, msg: Message, dst: usize) -> Result<()> {
+        let req = Request::builder()
+            .method(Method::POST)
+            .header("content-type", "application/json");
+
+        let req = req.uri(format!(
+            "{}/{}",
+            twin.address.trim_end_matches('/'),
+            uri_path
+        ));
+
+        let req = req
+            .body(Body::from(serde_json::to_vec(&msg).unwrap()))
+            .context(format!(
```

Comment:
It's really nice you started to use `context` to add more information to the error. But the problem with this now is that you will always create a new string even if there is no error. Because now you always calling `context` with the string returned by `format!` format then will always create a new string even if the call will not fail.

Instead `anyhow` also give u another method `with_context` witch accepts a closure (check example here) https://docs.rs/anyhow/latest/anyhow/trait.Context.html#example

```suggestion
            .with_context(|| format!("cannot construct request body for this id '{}'", dst))
```
the thing is now, the `format!` (and hence the creation of the new string that holds the error) will only happen if this closure is called which will happen only if an error happened.